### PR TITLE
DS 2682 Temporary KPI Dashboard

### PR DESCRIPTION
**Pull request by bochocki**
Ticket: https://mozilla-hub.atlassian.net/browse/DS-2682

The goal of this PR is to create a [temporary dashboard](https://mozilla.cloud.looker.com/dashboards/1132) that provides access to DAU and DAU 28MA (28-day Moving Average) for desktop and mobile.

## Changes
- `kpi/views/tmp_browsers_kpi_2023.view.lkml` implements business logic to calculate core metrics (DAU and DAU 28MA). Although implementing business logic in views is discouraged, this view is meant to be a temporary resource for stakeholders until we are ready to switch to the bigquery-etl version of the metrics. Currently, [there are some discrepancies between the query implemented here and the data available in the aggregated BQ tables](https://mozilla-hub.atlassian.net/browse/DS-2705).
- `kpi/explores/tmp_browsers_kpi_2023.explore.lkml` makes the `tmp_browsers_kpi_2023.view` available for visualization.
- Changes to imports in the following files seemed to be required to get the LookML error checker to pass, but maybe I was doing something wrong on my end:
  - `hubs/hubs.model.lkml`
  - `mach/mach.model.lkml`
  - `mozphab/mozphab.model.lkml`
  - `mozregression/mozregression.model.lkml`

image

Checklist for reviewer:

When adding a new derived dataset:
- [ ] Ensure that the data is not available already (fully or partially) and recommend extending an existing dataset in favor of creating new ones. Data may be available in [bigquery-etl repository](https://github.com/mozilla/bigquery-etl), [looker-hub](https://github.com/mozilla/looker-hub) or in [looker-spoke-default](https://github.com/mozilla/looker-spoke-default/tree/e1315853507fc1ac6e78d252d53dc8df5f5f322b).
- [ ] Avoid merging a PR that includes the logic of a [core metric](https://docs.telemetry.mozilla.org/metrics/index.html) or complex business logic. The recommendation is to implement core business logic in bigquery-etl. E.g. The [type of search](https://github.com/mozilla/bigquery-etl/blob/a3e59f90326816a2ecaaa3e9d5b57fe9552f7d70/sql/moz-fx-data-shared-prod/search_derived/mobile_search_clients_daily_v1/query.sql#L781) or the [calculation of DAU or visited URIs](https://github.com/mozilla/bigquery-etl/blob/9bca48821a8a0d40b1700cc14ecd8068d132ed06/sql/moz-fx-data-shared-prod/telemetry_derived/firefox_desktop_exact_mau28_by_dimensions_v1/query.sql).
- [ ] Avoid merging code in Looker Explores/Views that implement analysis with multiple lines of code or that will be likely replicated in the future. Instead, aim for extending an existing dataset to include the required logic, and use [Looker aggregates](https://cloud.google.com/looker/docs/aggregate_awareness) to facilitate the analysis.
- [ ] Avoid merging a PR with logic that requires validation and health checks. It is recommended to implement it in bigquery-etl for full test coverage and failure alerts.

### CODE REVIEW

File: kpi/views/tmp_browsers_kpi_2023.view.lkml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        'desktop' AS product,
+        COUNTIF(
+          active_hours_sum > 0 AND
+          COALESCE(
+            scalar_parent_browser_engagement_total_uri_count_normal_and_private_mode_sum,
+            scalar_parent_browser_engagement_total_uri_count_sum,
+            0
+          ) > 0
+        ) AS value
+      FROM `mozdata.telemetry.clients_daily`
```
Comment (lucia-vargas-a) [inline kpi/views/tmp_browsers_kpi_2023.view.lkml:17]: Do we want to use the the "shredded" numbers until July?
Before 2023-03-05 when shredder last ran, the numbers are not a 1:1 match with the active_users_aggregates tables. After that date, the [numbers between aggregates and clients_daily actually match exactly](https://docs.google.com/document/d/1J6VPYXin7Cf9aEItMadFLGooTmCaHhhTUSCaGTNcBlY/edit#heading=h.y2jiq1hzphef). The concern is that it generates confusion when we move to using the aggregates and DAU is different.

File: hubs/hubs.model.lkml [hunk 2]
```diff
@@ -2,4 +2,4 @@ connection: "telemetry"
 label: "Hubs"
 include: "views/*"
 include: "explores/*"
-include: "dashboards/*"
```
Comment (wwyc) [inline hubs/hubs.model.lkml:5]: A fix has been made and you should not need to comment out the unrelated code anymore.  Can you try pulling from production and try again?

File: kpi/views/tmp_browsers_kpi_2023.view.lkml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      WHERE submission_date >= '2022-12-16'
+      GROUP BY 1
+      ),
+
+      desktop_dau_28_ma AS (
```
Comment (wwyc) [inline kpi/views/tmp_browsers_kpi_2023.view.lkml:22]: moving averages can also be calculated using looker table calculations:  #https://help.looker.com/hc/en-us/articles/4420190523667-Moving-Averages-in-Table-Calculations-Community-

File: kpi/views/tmp_browsers_kpi_2023.view.lkml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        AVG(COUNTIF(
+          `mozfun`.bits28.active_in_range(days_seen_bits, 0, 1) AND
+          normalized_app_name IN ('Fenix', 'Firefox iOS')
+        )) OVER (ORDER BY submission_date ROWS BETWEEN 27 PRECEDING AND CURRENT ROW) AS value
+      FROM `mozdata.telemetry.unified_metrics`
+      WHERE submission_date >= DATE_SUB('2022-12-16', INTERVAL 27 DAY)
+      GROUP BY 1
+      ),
+
+      joined AS (
```
Comment (wwyc) [inline kpi/views/tmp_browsers_kpi_2023.view.lkml:65]: Instead of having 4 separate queries, have you considered combining desktop into one and mobile into one?  In that case, you would have two measures `dau` and `dau_28_ma` for each product instead of the `metric` and `value` measures you currently have.

File: hubs/hubs.model.lkml (hunk 2 above)
Comment (bochocki) [reply hubs/hubs.model.lkml]: Thanks! I undid these changes.

File: kpi/views/tmp_browsers_kpi_2023.view.lkml (hunk 3 above)
Comment (bochocki) [reply kpi/views/tmp_browsers_kpi_2023.view.lkml:22]: Thanks, I didn't know that! This looks like a really helpful pattern in general. In this specific case, I think we want to use pure SQL, even though it's much less elegant. One of the goals of this line of work is to use the same KPI definitions everywhere. In the short term, that will mean providing a SQL version of the query that can be used across the stack (see links below). In the long term, this logic will be implemented upstream in bq-etl so that we can just query the metric directly from some `active_users_aggregates` table. That will make the Metric Hub definitions simpler as well.

- Jira [DS-2680](https://mozilla-hub.atlassian.net/browse/DS-2680)
- [Metric Hub proposal](https://github.com/mozilla/metric-hub/pull/66)

File: kpi/views/tmp_browsers_kpi_2023.view.lkml (hunk 4 above)
Comment (bochocki) [reply kpi/views/tmp_browsers_kpi_2023.view.lkml:65]: This is a great idea, I'm refactoring to make this change now.

File: kpi/views/tmp_browsers_kpi_2023.view.lkml (hunk 3 above)
Comment (wwyc) [reply kpi/views/tmp_browsers_kpi_2023.view.lkml:22]: I agree.  We could eventually add the moving average column to the active users aggregates views. cc @lucia-vargas-a

File: kpi/views/tmp_browsers_kpi_2023.view.lkml (hunk 1 above)
Comment (bochocki) [reply kpi/views/tmp_browsers_kpi_2023.view.lkml:17]: We chatted offline and decided to communicated this difference when we roll over to `active_users_aggregates`.

The `clients_daily` table gets updated with "shredded" numbers automatically, but the `active_users_aggregates` tables do not account for shredded `client_ids` unless they're backfilled. This can actually be preferable, because it will mean that historic numbers won't change over time.

File: kpi/views/tmp_browsers_kpi_2023.view.lkml (hunk 4 above)
Comment (bochocki) [reply kpi/views/tmp_browsers_kpi_2023.view.lkml:45]: done!

**Review by wwyc — APPROVED** [review]
LGTM.
Once you have a final version of the looker dashboard, you can also consider having a [synced](https://github.com/mozilla/looker-spoke-default/blob/main/lookml_dashboards.yaml) lookml dashboard for better version control ([see documentation](https://cloud.google.com/looker/docs/building-lookml-dashboards)).

File: kpi/views/tmp_browsers_kpi_2023.view.lkml [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      FROM `mozdata.telemetry.clients_daily`
+      WHERE submission_date >= DATE_SUB('2022-12-16', INTERVAL 27 DAY)
+      GROUP BY 1
+      ),
+
+      mobile AS (
+      SELECT submission_date,
+        'mobile' AS product,
+        COUNTIF(
+          `mozfun`.bits28.active_in_range(days_seen_bits, 0, 1) AND
```
Comment (wwyc) [inline kpi/views/tmp_browsers_kpi_2023.view.lkml:33]: I am not familiar with the logic but `days_since_seen` column is also available in the `unified_metrics` table:  https://github.com/mozilla/bigquery-etl/blob/b4345961fde4b119dc6a1c9f361e0ddd9211fcb9/sql/moz-fx-data-shared-prod/telemetry_derived/active_users_aggregates_v1/query.sql#L65

Comment (irrationalagent) [reply kpi/views/tmp_browsers_kpi_2023.view.lkml:33]: +1  

`mozfun.bits28.active_in_range(days_seen_bits, 0, 1)` [should be strictly equivalent](https://docs.telemetry.mozilla.org/cookbooks/clients_last_seen_bits.html#bits28active_in_range) to `days_since_seen = 0` however the latter might be more transparent.

Comment (bochocki) [reply kpi/views/tmp_browsers_kpi_2023.view.lkml:33]: That's a good idea, y'all. For the sake of this temporary dashboard though, I'd like to stick with this definition because it's the one that's used in the 2023 [Mobile KPI Analysis](https://docs.google.com/document/d/1aNazhOEPexxiG_gAxIFJK-sHIHhEm0clXtvK1TtdAWw/edit#) -- the query is defined on [page 4 of this doc](https://drive.google.com/drive/folders/1WBRFAhiULyym9L08CWdFlCnhJmqnd_EB). We can keep our paper trail slightly simpler by sticking with this version until we update to use the new tables.

**Comment by bochocki** [comment]
> LGTM. Once you have a final version of the looker dashboard, you can also consider having a [synced](https://github.com/mozilla/looker-spoke-default/blob/main/lookml_dashboards.yaml) lookml dashboard for better version control ([see documentation](https://cloud.google.com/looker/docs/building-lookml-dashboards)).

Thanks Winnie, this is neat!
